dev/com.ibm.ws.repository.resolver/src/com/ibm/ws/repository/resolver/RepositoryResolver.java

```diff
@@ -223,331 +255,563 @@ public RepositoryResolver(Collection<ProductDefinition> installDefinition,
+                allDependenciesResolved &= populateMaxDistanceMap(maxDistanceMap, featureName, 1, new HashSet<ProvisioningFeatureDefinition>(), missingRequirements);
             }
         }
 
-        return resourcesToInstall;
+        if (!allDependenciesResolved) {
+            missingTopLevelRequirements.add(resource.getShortName());
+            this.missingRequirements.addAll(missingRequirements);
+        }
+
+        ArrayList<RepositoryResource> installList = new ArrayList<>();
+
+        installList.addAll(convertFeatureNamesToResources(maxDistanceMap.keySet()));
+        Collections.sort(installList, byMaxDistance(maxDistanceMap));
+
+        installList.add(resource);
+
+        return installList;
     }
 
     /**
-     * If the new install list is not empty this will remove any duplicates from it and add it to the resources to install set.
-     * 
-     * @param installList A list of resources to install
-     * @param resourcesToInstall A collection of all of the lists of resources to install
+     * Create a list of resources which should be installed in order to install the given featutre.
+     * <p>
+     * The install list consists of all the dependencies which are needed by {@code esa}, ordered so that each resource in the list comes after its dependencies.
+     *
+     * @param featureName the feature name (as provided to {@link #resolve(Collection)}) for which to create an install list
+     * @return the ordered list of resources to install
      */
-    private void addResourceList(List<RepositoryResource> installList, Collection<List<RepositoryResource>> resourcesToInstall) {
-        if (!installList.isEmpty()) {
-            /*
-             * Remove duplicates, easier doing this in a copy of the list so that we can make sure we always keep the first time something appears and don't have to worry
-             * about ConcurrentModificationExceptions
-             */
-            List<RepositoryResource> installListCopy = new ArrayList<RepositoryResource>(installList.size());
-            for (RepositoryResource massiveResource : installList) {
-                if (!installListCopy.contains(massiveResource)) {
-                    installListCopy.add(massiveResource);
-                }
+    List<RepositoryResource> createInstallList(String featureName) {
+        ProvisioningFeatureDefinition feature = resolverRepository.getFeature(featureName);
+        if (feature == null) {
+            // Feature missing
+            missingTopLevelRequirements.add(featureName);
+            // If we didn't find this feature in another product, we need to record it as missing
+            if (!requirementsFoundForOtherProducts.contains(featureName)) {
+                missingRequirements.add(new MissingRequirement(featureName, null));
             }
-            resourcesToInstall.add(installListCopy);
+            return Collections.emptyList();
         }
+
+        if (!(feature instanceof KernelResolverEsa)) {
```

Comment:
I'll check that the doc on `KernelResolverEsa` makes it clear that it represents a feature from the repository, rather than one from the installation.

Reply:
Done